File: Semana 5/actividad 3.3/main.cpp



#include <iostream>

class Contador {
	// Declaración de función amiga (friend)
	friend void modificarX(Contador& c, int nuevoValor);

public:
	Contador() : x(0) {}

	void mostrar() const {
		std::cout << "Valor de x: " << x << std::endl;
	}

private:
	int x;  // variable privada
};

// Definición de la función amiga
// Puede acceder directamente a x (privado) porque es friend
void modificarX(Contador& c, int nuevoValor) {
	c.x = nuevoValor;   // Acceso directo a miembro privado
}

int main() {
	Contador cont;
	cont.mostrar();          // Valor de x: 0
	modificarX(cont, 42);    // Modificación mediante friend
	cont.mostrar();          // Valor de x: 42

	// No se puede acceder a cont.x directamente desde main (error de compilación)
	// cont.x = 10;   // ERROR: 'int Contador::x' is private within this context

	return 0;
}

File: Semana 5/actividad 3.3/main.py


class Contador:
    def __init__(self):
        self.__x = 0   # Atributo "privado" (name mangling)
    
    def mostrar(self):
        print(f"Valor de x: {self.__x}")

# Función externa que modifica __x (simulando friend)
def modificar_x(contador, nuevo_valor):
    # Python no impide el acceso; el name mangling solo cambia el nombre
    contador._Contador__x = nuevo_valor   # Acceso directo al "privado"
    
    # También se podría simplemente asignar si fuera público o protegido
    # Pero respetamos la convención de que __x es privado

if __name__ == "__main__":
    c = Contador()
    c.mostrar()           # Valor de x: 0
    
    modificar_x(c, 99)    # Simula la función friend
    
    c.mostrar()           # Valor de x: 99
    
    # También es posible hacer esto (pero no es buena práctica)
    # c._Contador__x = 50

